src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -262,7 +386,14 @@ protected void CheckReentrancy()
         /// </summary>
         private void OnCollectionChanged(NotifyCollectionChangedAction action, object item, int index)
         {
-            OnCollectionChanged(new NotifyCollectionChangedEventArgs(action, item, index));
+            if (item is IList)
```

Comment:
This isn't going to work when someone is using a collection of collections (e.g. `ObservableCollection<IList>`) and you add a single `IList` to the collection (e.g. `c.Add(new[] { "Hello", "World" });`). The `CollectionChanged` event is going to incorrectly indicate that two items (`"Hello"` and `"World"`) were added to the collection, instead of correctly indicating a single item (a string array containing `"Hello"` and `"World"`) was added.

It also adds an unnecessary `if` check to every invocation that can be avoided.

Instead, consider adding new overloads of `OnCollectionChanged`:

```c#
// Existing
private void OnCollectionChanged(NotifyCollectionChangedAction action, object item, int index)
{
    OnCollectionChanged(new NotifyCollectionChangedEventArgs(action, item, index));
}

private void OnCollectionChanged(NotifyCollectionChangedAction action, object oldItem, object newItem, int index)
{
    OnCollectionChanged(new NotifyCollectionChangedEventArgs(action, newItem, oldItem, index));
}


// New
private void OnCollectionChanged(NotifyCollectionChangedAction action, IList items, int index)
{
    OnCollectionChanged(new NotifyCollectionChangedEventArgs(action, items, index));
}

private void OnCollectionChanged(NotifyCollectionChangedAction action, IList oldItems, IList newItems, int index)
{
    OnCollectionChanged(new NotifyCollectionChangedEventArgs(action, newItems, oldItems, index));
}
```

As an aside, these `OnCollectionChanged` helper overloads don't add a lot of value IMO... It might be worth  getting rid of them and just newing up an instance of `NotifyCollectionChangedEventArgs` and calling `OnCollectionChanged(NotifyCollectionChangedEventArgs e)` directly in all the call sites.

One thing to be aware of and something I found confusing about some of these helpers: the parameter order of the helper method is different than the order accepted by the `NotifyCollectionChangedEventArgs` constructor. For example, the following `OnCollectionChanged` method takes `oldItem` before `newItem`, but the `NotifyCollectionChangedEventArgs` constructor that it calls takes `newItem` before `oldItem`:

https://github.com/dotnet/corefx/blob/f52d7e6f69a35042489a192fec08edd318a134aa/src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs#L279-L282

Reply:
I added new overloads and updated the invocations so we are passing in IList as recommended.